tests/e2e/tests/stackdriver/stackdriver_test.go

```diff
@@ -0,0 +1,282 @@
+	"istio.io/istio/tests/util"
+)
+
+const (
+	fortioYaml  = "tests/e2e/tests/stackdriver/fortio-rules.yaml"
+	adapterYaml = "tests/e2e/tests/stackdriver/adapter.yaml"
+)
+
+type (
+	fortioTemplate struct {
+		FortioImage string
+	}
+	adapterTemplate struct {
+		Namespace string
+	}
+	testConfig struct {
+		*framework.CommonConfig
+	}
+)
+
+var (
+	tc      *testConfig
+	gcpProj *string
+)
+
+func (t *testConfig) Setup() error {
+	if !util.CheckPodsRunning(t.Kube.Namespace, t.Kube.KubeConfig) {
+		return fmt.Errorf("could not get all pods running")
+	}
+
+	filled, err := util.CreateAndFill(
+		t.Kube.TmpDir,
+		util.GetResourcePath(adapterYaml),
+		&adapterTemplate{
+			Namespace: t.Kube.Namespace,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	if err := util.KubeApply(t.Kube.Namespace, filled, t.Kube.KubeConfig); err != nil {
+		log.Errorf("Kubectl apply adapter yaml failed.")
+		return err
+	}
+
+	gateway, errGw := tc.Kube.IngressGateway()
+	if errGw != nil {
+		return errGw
+	}
+	if _, err := sendTrafficToCluster(gateway); err != nil {
+		return fmt.Errorf("generating HTTP traffic failed: %v", err)
+	}
+
+	// Push interval is 5s, wait for metrics to be written
+	time.Sleep(10 * time.Second)
+
+	return nil
+}
+
+func sendTrafficToCluster(gateway string) (*fhttp.HTTPRunnerResults, error) {
+	opts := fhttp.HTTPRunnerOptions{
+		RunnerOptions: periodic.RunnerOptions{
+			QPS:        10,
+			Exactly:    60,
+			NumThreads: 5,
+			Out:        os.Stderr,
+		},
+		HTTPOptions: fhttp.HTTPOptions{
+			URL: gateway + "/?status=418:10,520:15&size=1024:10,512:5",
+		},
+		AllowInitialErrors: true,
+	}
+	return fhttp.RunHTTPTest(&opts)
+}
+
+func (t *testConfig) Teardown() error {
+	return nil
+}
+
+func TestMain(m *testing.M) {
+	gcpProj = flag.String("gcp_proj", os.Getenv("GCP_PROJ"), "GCP Project ID, required")
+	flag.Parse()
+	check(framework.InitLogging(), "cannot setup logging")
+	check(setTestConfig(), "could not create TestConfig")
+	tc.Cleanup.RegisterCleanable(tc)
+	os.Exit(tc.RunTest(m))
+}
+
+func TestContextGraph(t *testing.T) {
+	// FIXME add test when read API is available.
+}
+
+func createMetricsService(ctx context.Context) (*monitoring.Service, error) {
+	hc, err := google.DefaultClient(ctx, monitoring.MonitoringReadScope)
+	if err != nil {
+		return nil, err
+	}
+	s, err := monitoring.New(hc)
+	if err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func TestMetrics(t *testing.T) {
+	const (
+		metricType     = "istio.io/service/server/request_count"
+		filterTemplate = `metric.type="%s" AND
+                      resource.labels.namespace_name = "%s" AND
+                      resource.type = "%s"`
+	)
+	projectResource := "projects/" + *gcpProj
+
+	ctx := context.Background()
+	s, err := createMetricsService(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	f := fmt.Sprintf(filterTemplate, metricType, tc.Kube.Namespace, "k8s_container")
+	t.Logf("Filter: %v", f)
+	startTime := time.Now().UTC().Add(time.Minute * -10)
```

Comment:
Curious: What happens when we run multiple instances of this test in the same project?  Does this imply that we should only ever run this as a nightly (or periodic) job?

Reply:
The namespace is randomly generated, all tests are looking for the namespace.

Reply:
Or does the `namespace_name` filter take care of that?